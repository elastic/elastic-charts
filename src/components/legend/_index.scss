@import 'variables';
@import 'legend';
@import 'legend_button';
@import 'legend_list';
@import 'legend_item';
