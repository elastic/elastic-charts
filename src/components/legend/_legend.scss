.echLegend {
  &--top,
  &--bottom {
    .echLegendList {
      display: grid;
      grid-column-gap: $echLegendColumnGap;
      grid-row-gap: $echLegendRowGap;
      margin-top: $echLegendRowGap;
      margin-bottom: $echLegendRowGap;

      @include internetExplorerOnly {
        display: flex;
        flex-direction: row;
        flex-wrap: wrap;
      }
    }
  }

  &--left,
  &--right {
    .echLegendList {
      flex-direction: column;
    }
  }

  &--top,
  &--left {
    order: 0;
  }

  &--bottom,
  &--right {
    order: 1;
  }

  &--debug {
    background: red;
    position: relative;
  }

  .echLegendListContainer {
    @include euiYScrollWithShadows;
    width: 100%;
    overflow-y: auto;
    overflow-x: hidden;
  }
}
