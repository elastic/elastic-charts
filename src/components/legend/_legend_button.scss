.echLegendButton {
  padding: $euiSizeXS;
  line-height: 1;
  opacity: 0.35;
  border-radius: $euiBorderRadius;
  background-color: $euiColorEmptyShade;
  position: absolute;
  bottom: 0;
  left: 0;
  transition:
    opacity $euiAnimSpeedFast $euiAnimSlightResistance,
    background-color $euiAnimSpeedFast $euiAnimSlightResistance;

  &:hover,
  &:focus {
    opacity: 1;
    background-color: $euiFocusBackgroundColor;
  }
}

.echLegendButton--isOpen {
  opacity: 1;
}
