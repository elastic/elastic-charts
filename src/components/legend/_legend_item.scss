$legendItemVerticalPadding: $echLegendRowGap / 2;

.echLegendItem {
  color: $euiTextColor;
  display: flex;
  flex-wrap: nowrap;
  justify-content: space-between;
  user-select: none;
  align-items: center;
  width: 100%;

  &:hover {
    .echLegendItem__label {
      text-decoration: underline;
    }
  }

  &__color {
    margin-right: $euiSizeXS;
  }

  &__visibility {
    margin-right: $euiSizeXS;

    &:hover {
      cursor: pointer;
    }
  }

  &__label {
    @include euiFontSizeXS;
    @include euiTextTruncate;
    flex: 1 1 auto;

    &:hover {
      cursor: pointer;
    }
  }

  &__label--hasClickListener {
    &:hover {
      cursor: pointer;
    }
  }

  &__extra {
    @include euiFontSizeXS;
    text-align: right;
    margin-left: $euiSizeXS;
    font-feature-settings: 'tnum';

    &--hidden {
      display: none;
    }
  }

  &--right,
  &--left {
    padding-top: $legendItemVerticalPadding;
    padding-bottom: $legendItemVerticalPadding;
  }

  @include internetExplorerOnly {
    &--bottom,
    &--top {
      width: $echLegendMaxWidth;
      margin-right: $euiSizeL;
    }

    padding-top: $legendItemVerticalPadding;
    padding-bottom: $legendItemVerticalPadding;
  }

  &--hidden {
    color: $euiColorDarkShade;
  }
}
