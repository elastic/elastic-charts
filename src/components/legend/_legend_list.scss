.echLegendList {
  @include euiScrollBar;
  height: 100%;
  max-width: 100%;
  width: 100%;
}
