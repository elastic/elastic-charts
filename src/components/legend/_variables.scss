$echLegendMaxWidth: 200px;
$echLegendRowGap: 8px;
$echLegendColumnGap: 24px;
