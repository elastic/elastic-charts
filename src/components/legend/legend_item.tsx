/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

import classNames from 'classnames';
import React, { Component, createRef, MouseEventHandler } from 'react';

import { LegendItem, LegendItemExtraValues } from '../../commons/legend';
import { SeriesIdentifier } from '../../commons/series_id';
import { LegendItemListener, BasicListener, LegendColorPicker, LegendAction } from '../../specs/settings';
import {
  clearTemporaryColors as clearTemporaryColorsAction,
  setTemporaryColor as setTemporaryColorAction,
  setPersistedColor as setPersistedColorAction,
} from '../../state/actions/colors';
import {
  onLegendItemOutAction,
  onLegendItemOverAction,
  onToggleDeselectSeriesAction,
} from '../../state/actions/legend';
import { Position, Color } from '../../utils/commons';
import { deepEqual } from '../../utils/fast_deep_equal';
import { Color as ItemColor } from './color';
import { renderExtra } from './extra';
import { Label as ItemLabel } from './label';
import { getExtra } from './utils';

/** @internal */
export const LEGEND_HIERARCHY_MARGIN = 10;

/** @internal */
export interface LegendItemProps {
  item: LegendItem;
  totalItems: number;
  position: Position;
  extraValues: Map<string, LegendItemExtraValues>;
  showExtra: boolean;
  colorPicker?: LegendColorPicker;
  action?: LegendAction;
  onClick?: LegendItemListener;
  onMouseOut?: BasicListener;
  onMouseOver?: LegendItemListener;
  mouseOutAction: typeof onLegendItemOutAction;
  mouseOverAction: typeof onLegendItemOverAction;
  clearTemporaryColorsAction: typeof clearTemporaryColorsAction;
  setTemporaryColorAction: typeof setTemporaryColorAction;
  setPersistedColorAction: typeof setPersistedColorAction;
  toggleDeselectSeriesAction: typeof onToggleDeselectSeriesAction;
}

/**
 * @internal
 * @param item
 * @param props
 */
export function renderLegendItem(
  item: LegendItem,
  props: Omit<LegendItemProps, 'item'>,
  totalItems: number,
  index: number,
) {
  const {
    seriesIdentifier: { key },
    childId,
  } = item;

  return (
    <LegendListItem
      key={`${key}-${childId}-${index}`}
      item={item}
      totalItems={totalItems}
      position={props.position}
      colorPicker={props.colorPicker}
      action={props.action}
      extraValues={props.extraValues}
      showExtra={props.showExtra}
      toggleDeselectSeriesAction={props.toggleDeselectSeriesAction}
      mouseOutAction={props.mouseOutAction}
      mouseOverAction={props.mouseOverAction}
      clearTemporaryColorsAction={props.clearTemporaryColorsAction}
      setTemporaryColorAction={props.setTemporaryColorAction}
      setPersistedColorAction={props.setPersistedColorAction}
      onMouseOver={props.onMouseOver}
      onMouseOut={props.onMouseOut}
      onClick={props.onClick}
    />
  );
}

interface LegendItemState {
  isOpen: boolean;
  actionActive: boolean;
}

/** @internal */
export class LegendListItem extends Component<LegendItemProps, LegendItemState> {
  static displayName = 'LegendItem';

  colorRef = createRef<HTMLDivElement>();
  state: LegendItemState = {
    isOpen: false,
    actionActive: false,
  };

  shouldComponentUpdate(nextProps: LegendItemProps, nextState: LegendItemState) {
    return !deepEqual(this.props, nextProps) || !deepEqual(this.state, nextState);
  }

  handleColorClick = (changable: boolean): MouseEventHandler | undefined =>
    changable
      ? (event) => {
          event.stopPropagation();
          this.toggleIsOpen();
        }
      : undefined;

  toggleIsOpen = () => {
    this.setState(({ isOpen }) => ({ isOpen: !isOpen }));
  };

  onLegendItemMouseOver = () => {
    const { onMouseOver, mouseOverAction, item } = this.props;
    // call the settings listener directly if available
    if (onMouseOver) {
      onMouseOver(item.seriesIdentifier);
    }
    mouseOverAction(item.seriesIdentifier.key);
  };

  onLegendItemMouseOut = () => {
    const { onMouseOut, mouseOutAction } = this.props;
    // call the settings listener directly if available
    if (onMouseOut) {
      onMouseOut();
    }
    mouseOutAction();
  };

  /**
   * Returns click function only if toggleable or click listern is provided
   */
  handleLabelClick = (legendItemId: SeriesIdentifier): MouseEventHandler | undefined => {
    const { item, onClick, toggleDeselectSeriesAction } = this.props;

    if (!item.isToggleable && !onClick) {
      return;
    }

    return ({ shiftKey }) => {
      if (onClick) {
        onClick(legendItemId);
      }

      if (item.isToggleable) {
        toggleDeselectSeriesAction(legendItemId, shiftKey);
      }
    };
  };

  renderColorPicker() {
    const {
      colorPicker: ColorPicker,
      item,
      clearTemporaryColorsAction,
      setTemporaryColorAction,
      setPersistedColorAction,
    } = this.props;
    const { seriesIdentifier, color } = item;

    const handleClose = () => {
      setPersistedColorAction(seriesIdentifier.key, color);
      clearTemporaryColorsAction();
      this.toggleIsOpen();
    };
    if (ColorPicker && this.state.isOpen && this.colorRef.current) {
      return (
        <ColorPicker
          anchor={this.colorRef.current}
          color={color}
          onClose={handleClose}
          onChange={(color: Color) => setTemporaryColorAction(seriesIdentifier.key, color)}
          seriesIdentifier={seriesIdentifier}
        />
      );
    }
  }

  render() {
    const { extraValues, item, showExtra, colorPicker, position, totalItems, action: Action } = this.props;
    const { color, isSeriesHidden, isItemHidden, seriesIdentifier, label } = item;
    const itemClassNames = classNames('echLegendItem', `echLegendItem--${position}`, {
      'echLegendItem--hidden': isSeriesHidden,
      'echLegendItem__extra--hidden': isItemHidden,
    });
    const hasColorPicker = Boolean(colorPicker);
    const extra = getExtra(extraValues, item, totalItems);
    const style = item.depth
      ? {
          marginLeft: LEGEND_HIERARCHY_MARGIN * (item.depth ?? 0),
        }
      : undefined;
    return (
      <>
        <li
          className={itemClassNames}
          onMouseEnter={this.onLegendItemMouseOver}
          onMouseLeave={this.onLegendItemMouseOut}
          style={style}
          data-ech-series-name={label}
        >
          <ItemColor
            ref={this.colorRef}
            color={color}
            isSeriesHidden={isSeriesHidden}
            hasColorPicker={hasColorPicker}
            onClick={this.handleColorClick(hasColorPicker)}
          />
          <ItemLabel label={label} onClick={this.handleLabelClick(seriesIdentifier)} />
          {showExtra && extra != null && renderExtra(extra, isSeriesHidden)}
          {Action && (
            <div className="echLegendItem__action">
              <Action series={seriesIdentifier} color={color} label={label} />
            </div>
          )}
        </li>
        {this.renderColorPicker()}
      </>
    );
  }
}
