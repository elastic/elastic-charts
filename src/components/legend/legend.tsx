import classNames from 'classnames';
import { inject, observer } from 'mobx-react';
import React, { createRef } from 'react';
import { isVerticalAxis, isHorizontalAxis } from '../../chart_types/xy_chart/utils/axis_utils';
import { LegendItem as SeriesLegendItem } from '../../chart_types/xy_chart/legend/legend';
import { ChartStore } from '../../chart_types/xy_chart/store/chart_state';
import { Position } from '../../chart_types/xy_chart/utils/specs';
import { LegendItem } from './legend_item';
import { Theme } from '../../utils/themes/theme';

interface LegendProps {
  chartStore?: ChartStore; // FIX until we find a better way on ts mobx
}

interface LegendState {
  width?: number;
}

interface LegendStyle {
  maxHeight?: string;
  maxWidth?: string;
  width?: string;
}

interface LegendListStyle {
  paddingTop?: number | string;
  paddingBottom?: number | string;
  paddingLeft?: number | string;
  paddingRight?: number | string;
  gridTemplateColumns?: string;
}

class LegendComponent extends React.Component<LegendProps, LegendState> {
  static displayName = 'Legend';

  state = {
    width: undefined,
  };

  private echLegend = createRef<HTMLDivElement>();

  componentDidUpdate() {
    const { chartInitialized, chartTheme, legendPosition } = this.props.chartStore!;
    if (
      this.echLegend.current &&
      isVerticalAxis(legendPosition.get()) &&
      this.state.width === undefined &&
      !chartInitialized.get()
    ) {
      const buffer = chartTheme.legend.spacingBuffer;

      this.setState({
        width: this.echLegend.current.offsetWidth + buffer,
      });
    }
  }

  render() {
    const {
      legendInitialized,
      chartInitialized,
      legendItems,
      legendPosition,
      showLegend,
      debug,
      chartTheme,
    } = this.props.chartStore!;
    const position = legendPosition.get();

    if (!showLegend.get() || !legendInitialized.get() || legendItems.size === 0) {
      return null;
    }

    const legendContainerStyle = this.getLegendStyle(position, chartTheme);
    const legendListStyle = this.getLegendListStyle(position, chartTheme);
    const legendClasses = classNames('echLegend', `echLegend--${position}`, {
      'echLegend--debug': debug,
      invisible: !chartInitialized.get(),
    });

    return (
      <div ref={this.echLegend} className={legendClasses}>
        <div style={legendContainerStyle} className="echLegendListContainer">
          <div style={legendListStyle} className="echLegendList">
            {[...legendItems.values()].map(this.renderLegendElement)}
          </div>
        </div>
      </div>
    );
  }

  getLegendListStyle = (position: Position, { chartMargins, legend }: Theme): LegendListStyle => {
    const { top: paddingTop, bottom: paddingBottom, left: paddingLeft, right: paddingRight } = chartMargins;

    if (isHorizontalAxis(position)) {
      return {
        paddingLeft,
        paddingRight,
        gridTemplateColumns: `repeat(auto-fill, minmax(${legend.verticalWidth}px, 1fr))`,
      };
    }

    return {
      paddingTop,
      paddingBottom,
    };
  };

  getLegendStyle = (position: Position, { legend }: Theme): LegendStyle => {
    if (isVerticalAxis(position)) {
      if (this.state.width !== undefined) {
        const threshold = Math.min(this.state.width!, legend.verticalWidth);
        const width = `${threshold}px`;

        return {
          width,
          maxWidth: width,
        };
      }

      return {
        maxWidth: `${legend.verticalWidth}px`,
      };
    }

    return {
      maxHeight: `${legend.horizontalHeight}px`,
    };
  };

  onLegendItemMouseover = (legendItemKey: string) => () => {
    this.props.chartStore!.onLegendItemOver(legendItemKey);
  };

  onLegendItemMouseout = () => {
    this.props.chartStore!.onLegendItemOut();
  };

  private renderLegendElement = (item: SeriesLegendItem) => {
    const { key, displayValue } = item;
    const { legendPosition, legendItemTooltipValues, isCursorOnChart } = this.props.chartStore!;
    const tooltipValues = legendItemTooltipValues.get();
    let tooltipValue;

    if (tooltipValues && tooltipValues.get(key)) {
      tooltipValue = tooltipValues.get(key);
    }

    const newDisplayValue = tooltipValue != null ? tooltipValue : '';

    return (
      <LegendItem
        {...item}
        key={key}
        legendItemKey={key}
        legendPosition={legendPosition.get()}
        displayValue={isCursorOnChart.get() ? newDisplayValue : displayValue.formatted}
        onMouseEnter={this.onLegendItemMouseover(key)}
        onMouseLeave={this.onLegendItemMouseout}
      />
    );
  };
}

export const Legend = inject('chartStore')(observer(LegendComponent));
