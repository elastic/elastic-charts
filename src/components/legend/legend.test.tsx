import React from 'react';
import { mount } from 'enzyme';
import { Chart } from '../chart';
import { Settings, BarSeries } from '../../specs';
import { ScaleType } from '../../scales';
import { DataGenerator } from '../../utils/data_generators/data_generator';
import { Legend } from './legend';
import { LegendListItem } from './legend_item';

describe('Legend', () => {
  it('shall render the all the series names', () => {
    const wrapper = mount(
      <Chart>
        <Settings showLegend />
        <BarSeries
          id="areas"
          name="area"
          xScaleType={ScaleType.Linear}
          yScaleType={ScaleType.Linear}
          xAccessor={0}
          yAccessors={[1]}
          splitSeriesAccessors={[2]}
          data={[
            [0, 123, 'group0'],
            [0, 123, 'group1'],
            [0, 123, 'group2'],
            [0, 123, 'group3'],
          ]}
        />
      </Chart>,
    );
    const legendWrapper = wrapper.find(Legend);
    expect(legendWrapper.exists).toBeTruthy();
    const legendItems = legendWrapper.find(LegendListItem);
    expect(legendItems.exists).toBeTruthy();
    expect(legendItems).toHaveLength(4);
    legendItems.forEach((legendItem, i) => {
      // the legend item shows also the value as default parameter
      expect(legendItem.text()).toBe(`group${i}123`);
    });
  });
  it('shall render the all the series names without the data value', () => {
    const wrapper = mount(
      <Chart>
        <Settings showLegend showLegendDisplayValue={false} />
        <BarSeries
          id="areas"
          name="area"
          xScaleType={ScaleType.Linear}
          yScaleType={ScaleType.Linear}
          xAccessor={0}
          yAccessors={[1]}
          splitSeriesAccessors={[2]}
          data={[
            [0, 123, 'group0'],
            [0, 123, 'group1'],
            [0, 123, 'group2'],
            [0, 123, 'group3'],
          ]}
        />
      </Chart>,
    );
    const legendWrapper = wrapper.find(Legend);
    expect(legendWrapper.exists).toBeTruthy();
    const legendItems = legendWrapper.find(LegendListItem);
    expect(legendItems.exists).toBeTruthy();
    expect(legendItems).toHaveLength(4);
    legendItems.forEach((legendItem, i) => {
      // the legend item shows also the value as default parameter
      expect(legendItem.text()).toBe(`group${i}`);
    });
  });
  it('shall call the over and out listeners for every list item', () => {
    const onLegendItemOver = jest.fn();
    const onLegendItemOut = jest.fn();
    const dg = new DataGenerator();
    const numberOfSeries = 4;
    const data = dg.generateGroupedSeries(10, numberOfSeries, 'split');
    const wrapper = mount(
      <Chart>
        <Settings showLegend onLegendItemOver={onLegendItemOver} onLegendItemOut={onLegendItemOut} />
        <BarSeries
          id="areas"
          xScaleType={ScaleType.Linear}
          yScaleType={ScaleType.Linear}
          xAccessor={'x'}
          yAccessors={['y']}
          splitSeriesAccessors={['g']}
          data={data}
        />
      </Chart>,
    );
    const legendWrapper = wrapper.find(Legend);
    expect(legendWrapper.exists).toBeTruthy();
    const legendItems = legendWrapper.find(LegendListItem);
    expect(legendItems.exists).toBeTruthy();
    legendItems.forEach((legendItem, i) => {
      legendItem.simulate('mouseenter');
      expect(onLegendItemOver).toBeCalledTimes(i + 1);
      legendItem.simulate('mouseleave');
      expect(onLegendItemOut).toBeCalledTimes(i + 1);
    });
  });
  it('shall call click listener for every list item', () => {
    const onLegendItemClick = jest.fn();
    const dg = new DataGenerator();
    const numberOfSeries = 4;
    const data = dg.generateGroupedSeries(10, numberOfSeries, 'split');
    const wrapper = mount(
      <Chart>
        <Settings showLegend onLegendItemClick={onLegendItemClick} />
        <BarSeries
          id="areas"
          xScaleType={ScaleType.Linear}
          yScaleType={ScaleType.Linear}
          xAccessor={'x'}
          yAccessors={['y']}
          splitSeriesAccessors={['g']}
          data={data}
        />
      </Chart>,
    );
    const legendWrapper = wrapper.find(Legend);
    expect(legendWrapper.exists).toBeTruthy();
    const legendItems = legendWrapper.find(LegendListItem);
    expect(legendItems.exists).toBeTruthy();
    expect(legendItems).toHaveLength(4);
    legendItems.forEach((legendItem, i) => {
      // the click is only enabled on the title
      legendItem.find('.echLegendItem__title').simulate('click');
      expect(onLegendItemClick).toBeCalledTimes(i + 1);
    });
  });
});
