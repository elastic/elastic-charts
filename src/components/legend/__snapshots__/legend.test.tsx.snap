// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Legend #legendColorPicker should match snapshot after onChange is called 1`] = `"<li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splita\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"#0c7b93\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splita\\">splita</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitb\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"#0c7b93\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitb\\">splitb</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitc\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"#0c7b93\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitc\\">splitc</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitd\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"#0c7b93\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitd\\">splitd</div></li>"`;

exports[`Legend #legendColorPicker should match snapshot after onClose is called 1`] = `"<li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splita\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splita\\">splita</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitb\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitb\\">splitb</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitc\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitc\\">splitc</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitd\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitd\\">splitd</div></li>"`;

exports[`Legend #legendColorPicker should render colorPicker when color is clicked 1`] = `"<div id=\\"colorPicker\\"><span>Custom Color Picker</span><button id=\\"change\\" type=\\"button\\">#0c7b93</button><button id=\\"close\\" type=\\"button\\">close</button></div>"`;

exports[`Legend #legendColorPicker should render colorPicker when color is clicked 2`] = `"<li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splita\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splita\\">splita</div></li><div id=\\"colorPicker\\"><span>Custom Color Picker</span><button id=\\"change\\" type=\\"button\\">#0c7b93</button><button id=\\"close\\" type=\\"button\\">close</button></div><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitb\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitb\\">splitb</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitc\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitc\\">splitc</div></li><li class=\\"echLegendItem echLegendItem--right\\" data-ech-series-name=\\"splitd\\"><div class=\\"echLegendItem__color echLegendItem__color--changable\\" aria-label=\\"series color\\" title=\\"change series color\\"><div><svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"16\\" height=\\"16\\" class=\\"echIcon\\" color=\\"red\\" focusable=\\"false\\"><defs><circle id=\\"dot-a\\" cx=\\"8\\" cy=\\"8\\" r=\\"4\\"></circle></defs><g><use xlink:href=\\"#dot-a\\"></use></g></svg></div></div><div class=\\"echLegendItem__label echLegendItem__label--clickable\\" title=\\"splitd\\">splitd</div></li>"`;
